plugins/inventory/cloudstack.py

```diff
@@ -0,0 +1,232 @@
+=============================================
+
+Generates Ansible inventory from CloudStack. Configuration is read from
+'cloudstack.ini'. If you need to pass the project, write a simple wrapper
+script, e.g. project_cloudstack.sh:
+
+  #!/bin/bash
+  cloudstack.py --project <your_project> $@
+
+
+When run against a specific host, this script returns the following attributes
+based on the data obtained from CloudStack API:
+
+  "web01": {
+    "cpu_number": 2,
+    "nic": [
+      {
+        "ip": "10.102.76.98",
+        "mac": "02:00:50:99:00:01",
+        "type": "Isolated",
+        "netmask": "255.255.255.0",
+        "gateway": "10.102.76.1"
+      },
+      {
+        "ip": "10.102.138.63",
+        "mac": "06:b7:5a:00:14:84",
+        "type": "Shared",
+        "netmask": "255.255.255.0",
+        "gateway": "10.102.138.1"
+      }
+    ],
+    "default_ip": "10.102.76.98",
+    "zone": "ZUERICH",
+    "created": "2014-07-02T07:53:50+0200",
+    "hypervisor": "VMware",
+    "memory": 2048,
+    "state": "Running",
+    "tags": [],
+    "cpu_speed": 1800,
+    "affinity_group": [],
+    "service_offering": "Small",
+    "cpu_used": "62%"
+  }
+
+
+usage: cloudstack.py [--list] [--host HOST] [--project PROJECT]
+"""
+
+import os
+import sys
+import argparse
+
+try:
+    import json
+except:
+    import simplejson as json
+
+
+try:
+    from cs import CloudStack, CloudStackException, read_config
+except ImportError:
+    print >> sys.stderr, "Error: CloudStack library must be installed: pip install cs."
+    sys.exit(1)
+
+
+class CloudStackInventory(object):
+    def __init__(self):
+
+        parser = argparse.ArgumentParser()
+        parser.add_argument('--host')
+        parser.add_argument('--list', action='store_true')
+        parser.add_argument('--project')
+
+        options = parser.parse_args()
+        try:
+            self.cs = CloudStack(**read_config())
+        except CloudStackException, e:
+            print >> sys.stderr, "Error: Could not connect to CloudStack API"
+
+        project_id = ''
+        if options.project:
+            project_id = self.get_project_id(options.project)
+
+        if options.host:
+            data = self.get_host(options.host)
+            print json.dumps(data, indent=2)
+
+        elif options.list:
+            data = self.get_list()
+            print json.dumps(data, indent=2)
+        else:
+            print >> sys.stderr, "usage: --list | --host <hostname> [--project <project>]"
+            sys.exit(1)
+
+
+    def get_project_id(self, project):
+        projects = self.cs.listProjects()
+        if projects:
+            for p in projects['project']:
+                if p['name'] == project or p['id'] == project:
+                    return p['id']
+        print >> sys.stderr, "Error: Project %s not found." % project
+        sys.exit(1)
+
+
+    def get_host(self, name, project_id=''):
+        hosts = self.cs.listVirtualMachines(projectid=project_id)
```

Comment:
This doesn't work when list of projects is empty (example: exoscale.ch). It results with following error:

`Traceback (most recent call last):
  File "./cloudstack.py", line 232, in <module>
    CloudStackInventory()
  File "./cloudstack.py", line 113, in __init__
    data = self.get_list()
  File "./cloudstack.py", line 178, in get_list
    groups = self.cs.listInstanceGroups(projectid=project_id)
  File "/Library/Python/2.7/site-packages/cs/client.py", line 105, in handler
    return self._request(command, **kwargs)
  File "/Library/Python/2.7/site-packages/cs/client.py", line 157, in _request
    response.status_code), response, data)
cs.client.CloudStackException: ('HTTP 431 response from CloudStack', <Response [431]>, {u'errorcode': 431, u'uuidList': [], u'cserrorcode': 9999, u'errortext': u'Unable to execute API command listinstancegroups due to invalid value. Invalid parameter projectid value= due to incorrect long value format, or entity does not exist or due to incorrect parameter annotation for the field in api cmd class.'})`

It works, when I remove `projectid=project_id` form the API calls.